### PR TITLE
Using GCC(<5.0) compile ONNX may cause some error.( #3004)

### PR DESCRIPTION
**Pull request by 462630221**
Modify tensor.h. GCC(<5.0) use Copy-On-Write(COW) for std::string. It may cause shallow copy. So I add copy constructor that force deep copy in struct tensor, to avoid shallow copy.

### CODE REVIEW

File: onnx/common/tensor.h [hunk 1]
```diff
@@ -50,6 +50,51 @@ struct Tensor final {
...
+    // Deep copy. Avoid copy on write when using gcc<5.0
+    string_data_.resize(other.string_data_.size());
+    for(unsigned int i=0; i<other.string_data_.size(); ++i) {
+      string_data_[i] = std::string( other.string_data_[i].data(), other.string_data_[i].size() );
+    }
+    name_ = std::string(other.name_.data(), other.name_.size());
+    raw_data_ = std::string(other.raw_data_.data(), other.raw_data_.size());
+  }
+
+  Tensor& operator = (const Tensor &ohther){
```
Comment (daquexian) [inline onnx/common/tensor.h:74]: typo, ohther->other

File: onnx/common/tensor.h [hunk 2]
```diff
@@ -50,6 +50,51 @@ struct Tensor final {
...
+    if(this != &other){
+        is_segment_ = other.is_segment_;
```
Comment (daquexian) [inline onnx/common/tensor.h:76]: inconsistent indent

File: onnx/common/tensor.h [hunk 3]
```diff
@@ -50,6 +50,51 @@ struct Tensor final {
...
+        has_name_ = other.has_name_;
+        elem_type_ = other.elem_type_;
+        sizes_ = other.sizes_;
+        float_data_ = other.float_data_;
+        double_data_ = other.double_data_;
+        int32_data_ = other.int32_data_;
+        int64_data_ = other.int64_data_;
+        is_raw_data_ = other.is_raw_data_;
+        // Deep copy. Avoid copy on write when using gcc<5.0
+        string_data_.resize(other.string_data_.size());
```
Comment (daquexian) [inline onnx/common/tensor.h:88]: This copy assignment implementation doesn't guarantee strong exception safety. Maybe [copy-and-swap](https://en.wikibooks.org/wiki/More_C%2B%2B_Idioms/Copy-and-swap) like [here](https://github.com/onnx/onnx/issues/3004#issuecomment-692487726) is a better implementation (It's just an advice).

Comment (daquexian) [reply onnx/common/tensor.h:88]: > 1 for Inconsistent definition of copy constructor and assignment ('Rule of Two')

LGTM analysis also complains about this

File: onnx/common/tensor.h [hunk 4]
```diff
@@ -50,6 +50,51 @@ struct Tensor final {
...
+  , segment_begin_(other.segment_begin_)
+  , segment_end_(other.segment_end_)
+  , has_name_(other.has_name_)
+  , elem_type_(other.elem_type_)
+  , sizes_(other.sizes_)
+  , float_data_(other.float_data_)
+  , double_data_(other.double_data_)
+  , int32_data_(other.int32_data_)
+  , int64_data_(other.int64_data_)
+  , is_raw_data_(other.is_raw_data_) {
```
Comment (daquexian) [inline onnx/common/tensor.h:64]: uint64_data_ is missing

**Review by daquexian — APPROVED** [review]
I have verified that this patch fixes #3004

**Comment by daquexian** [comment]
@462630221 Could you please sign the CLA and merge the latest master branch?

**Review by linkerzhang — APPROVED** [review]
lgtm.

**Pull request merged**